Source/WebCore/style/AnchorPositionEvaluator.cpp

```diff
@@ -741,9 +741,59 @@ static bool firstChildPrecedesSecondChild(const RenderObject* firstChild, const
     return false;
 }
 
+// Given an anchor element and its anchor names, locate the closest ancestor element
+// that establishes an anchor scope affecting this anchor element, and return the pointer
+// to such element. If no ancestor establishes an anchor scope affecting this anchor,
+// returns nullptr.
+static CheckedPtr<Element> anchorScopeForAnchorElement(const Element& anchorElement, const Vector<ScopedName>& anchorNames)
+{
+    // Precondition: anchorElement is an anchor, which has at least one anchor name.
+    ASSERT(!anchorNames.isEmpty());
+
+    // Traverse up the element tree through each ancestor.
+    for (CheckedPtr currentAncestor = anchorElement.parentElement(); currentAncestor; currentAncestor = currentAncestor->parentElement()) {
```

Comment:
I think this should be using `parentElementInComposedTree()`. The spec text says

> Like most operations in CSS besides selector matching, features in this specification operate over the [flattened element tree](https://drafts.csswg.org/css-scoping-1/#flat-tree).

(webkit uses "composed tree" for "flat tree")